File: src/main/java/com/tms/csp/parse/TokenStream.java
package com.tms.csp.parse;

import java.util.ArrayList;
import java.util.List;

import static com.google.common.base.Preconditions.checkState;

public class TokenStream {

    private final GlobalTokenizer tokenizer;

    private final Stream stream;

    Token buffer;

    public TokenStream(GlobalTokenizer tokenizer, Stream stream) {
        this.tokenizer = tokenizer;
        this.stream = stream;
    }

    public TokenStream(GlobalTokenizer tokenizer, String inputText) {
        this(tokenizer, new Stream(inputText));
    }

    public List<Token> readAllTokens() {
        ArrayList<Token> a = new ArrayList<Token>();
        while (true) {
            if (stream.isEol()) {
                break;
            }
            Token token = take();
            a.add(token);
        }
        return a;

    }

    public void printTokens() {
        Token t = take();
        while (t != null) {
            System.err.println(t);
        }

    }

    public Token peek() {
        if (this.buffer == null) {
            this.buffer = tokenizer.takeNextToken(stream);
        }
        return this.buffer;
    }

//    public Token take() {
//        if (tokens.isEmpty()) {
//            return null;
//        } else {
//            return tokens.remove(0);
//        }
//    }

//    public Token takeNextTokenFromStream() {
//        return tokenizer.takeNextToken(stream);
//    }

    public Token take() {
        if (this.buffer != null) {
            Token tmp = buffer;
            this.buffer = null;
            return tmp;
        } else {
            return tokenizer.takeNextToken(stream);
        }
    }

    public GlobalTokenizer getTokenizer() {
        return tokenizer;
    }

    public Stream getStream() {
        return stream;
    }


    public void consumeAndCheck(char c) {
        Token t = consume();
        char cc = t.getChar();
        checkState(cc == c, "check failed - expected[" + c + "]   got[" + cc + "]");
    }

    public Token consume() {
        return take();
    }

    public boolean isEmpty() {
        return stream.isEol() && buffer == null;
    }


}

File: src/main/java/com/tms/csp/parse/TokenStream.kt
package com.tms.csp.parse

import java.util.ArrayList

import com.google.common.base.Preconditions.checkState

class TokenStream(val tokenizer: GlobalTokenizer, val stream: Stream) {

    internal var _buffer: Token? = null

    val isEmpty: Boolean
        get() = stream.isEol && _buffer == null

    constructor(tokenizer: GlobalTokenizer, inputText: String) : this(tokenizer, Stream(inputText)) {}

    fun readAllTokens(): List<Token> {
        val a = ArrayList<Token>()
        while (true) {
            if (stream.isEol) {
                break
            }
            val token:Token = take()!!
            a.add(token)
        }
        return a

    }

    fun printTokens() {
        val t = take()
        while (t != null) {
            System.err.println(t)
        }

    }

    fun peek(): Token {
        if (this._buffer == null) {
            this._buffer = tokenizer.takeNextToken(stream)
        }
        return this._buffer!!
    }

    //    public Token take() {
    //        if (tokens.isEmpty()) {
    //            return null;
    //        } else {
    //            return tokens.remove(0);
    //        }
    //    }

    //    public Token takeNextTokenFromStream() {
    //        return tokenizer.takeNextToken(stream);
    //    }

    fun take(): Token? {
        if (this._buffer != null) {
            val tmp = _buffer
            this._buffer = null
            return tmp
        } else {
            return tokenizer.takeNextToken(stream)
        }
    }


    fun consumeAndCheck(c: Char) {
        val t = consume()
        val cc = t!!.char!!
        checkState(cc == c, "check failed - expected[$c]   got[$cc]")
    }

    fun consume(): Token? {
        return take()
    }


}
